src/compressor/QatAccel.cc

```diff
@@ -48,22 +53,111 @@ bool QatAccel::init(const std::string &alg) {
   rc = qzSetDefaults(&params);
   if (rc != QZ_OK)
       return false;
+  return true;
+}
 
+static bool SetupSession(QatAccel::session_ptr &session, QzSessionParams_T &params) {
+  int rc;
   rc = qzInit(session.get(), QZ_SW_BACKUP_DEFAULT);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW)
+  if (rc != QZ_OK && rc != QZ_DUPLICATE)
     return false;
-
   rc = qzSetupSession(session.get(), &params);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW ) {
-    qzTeardownSession(session.get());
-    qzClose(session.get());
+  if (rc != QZ_OK) {
+    return false;
+  }
+  return true;
+}
+
+// put the session back to the session pool in a RAII manner
+struct cached_session_t {
+  cached_session_t(QatAccel* accel)
+    : accel{accel} {
+    std::unique_lock lock{accel->mutex};
+    if (!accel->sessions.empty()) {
+      session = std::move(accel->sessions.back());
+      accel->sessions.pop_back();
+      return;
+    }
+
+    // If there are no available session to use, we try allocate a new
+    // session. When we failed to allocate a new session, we block on here,
+    // waiting a available session;
+    QzSessionParams_T params = {(QzHuffmanHdr_T)0,};
+    QatAccel::session_ptr s{new struct QzSession_S()};
+    memset(s.get(), 0, sizeof(struct QzSession_S));
+    if (getQzParams(accel->alg, params) && SetupSession(s, params)) {
+      session = std::move(s);
+    } else {
+      accel->cond.wait(lock, [accel] { return !accel->sessions.empty(); });
+      session = std::move(accel->sessions.back());
+      accel->sessions.pop_back();
+    }
+  }
+  ~cached_session_t() {
+    std::scoped_lock lock{accel->mutex};
+    // if sessions.size() is large and equal than qat_compressor_session_max_number, we release current session.
```

Comment:
```suggestion
    // if the cache size is still under its upper bound, the current session is put into
    // accel->sessions. otherwise it's released right away.
```

in general, would be better to put "is greater or equal to". but i think it'd be more readable to emphasis
that we are trying to resuse the session by caching it as long as the cache size is under the configured upper bound.

Reply:
I agree your view and keep it.